pkg/network/usm/sharedlibraries/ebpf.go

```diff
@@ -126,19 +181,182 @@ func (e *EbpfProgram) Init() error {
+	}
+
+	if err := e.start(); err != nil {
+		return fmt.Errorf("cannot start manager: %w", err)
+	}
+
+	e.isInitialized = true
+	return nil
 }
 
-// Stop stops the eBPF program
+func (e *EbpfProgram) start() error {
+	err := e.Manager.Start()
+	if err != nil {
+		return fmt.Errorf("cannot init manager: %w", err)
+	}
+
+	for libset, perfHandler := range e.perfHandlers {
+		e.wg.Add(1)
+		go func(ls Libset, ph *ddebpf.PerfHandler) {
+			defer e.wg.Done()
+
+			dataChan := ph.DataChannel()
+			lostChan := ph.LostChannel()
+			for {
+				select {
+				case <-e.done[ls]:
+					return
+				case event, ok := <-dataChan:
+					if !ok {
+						return
+					}
+					e.handleEvent(&event, ls)
+				case <-lostChan:
+					// Nothing to do in this case
+					break
+				}
+			}
+		}(libset, perfHandler)
+	}
+
+	return nil
+}
+
+func (e *EbpfProgram) handleEvent(event *ebpf.DataEvent, libset Libset) {
+	e.callbacksMutex.RLock()
+	defer func() {
+		event.Done()
+		e.callbacksMutex.RUnlock()
+	}()
+
+	libpath := ToLibPath(event.Data)
+	for callback := range e.callbacks[libset] {
+		// Not using a callback runner for now, as we don't have a lot of callbacks
+		(*callback)(libpath)
+	}
+}
+
+// CheckLibsetsEnabled checks if the eBPF program is enabled for the given libsets, returns an error if not
+// with the libsets that are not enabled
+func (e *EbpfProgram) CheckLibsetsEnabled(libsets ...Libset) error {
+	e.initMutex.Lock()
+	defer e.initMutex.Unlock()
+
+	var err error
+	for _, libset := range libsets {
+		if _, ok := e.enabledLibsets[libset]; !ok {
+			err = multierror.Append(err, fmt.Errorf("libset %s is not enabled", libset))
+		}
+	}
+
+	return err
+}
+
+// Subscribe subscribes to the shared libraries events for the given libsets, returns the function
+// to call to unsubscribe and an error if the libsets are not enabled
+func (e *EbpfProgram) Subscribe(callback LibraryCallback, libsets ...Libset) (func(), error) {
+	if err := e.CheckLibsetsEnabled(libsets...); err != nil {
+		return nil, err
+	}
+
+	e.callbacksMutex.Lock()
+	defer e.callbacksMutex.Unlock()
+
+	for _, libset := range libsets {
+		if _, ok := e.callbacks[libset]; !ok {
+			e.callbacks[libset] = make(map[*LibraryCallback]struct{})
+		}
+		e.callbacks[libset][&callback] = struct{}{}
+	}
+
+	// UnSubscribe()
+	return func() {
+		e.callbacksMutex.Lock()
+		defer e.callbacksMutex.Unlock()
+		for _, libset := range libsets {
+			delete(e.callbacks[libset], &callback)
+		}
+	}, nil
+}
+
+// Stop stops the eBPF program if the refcount reaches 0
 func (e *EbpfProgram) Stop() {
+	if e.refcount.Dec() != 0 {
+		if e.refcount.Load() < 0 {
+			e.refcount.Swap(0)
+		}
+		return
+	}
+
+	log.Info("shared libraries monitor stopping due to a refcount of 0")
+
+	e.stopImpl()
+
+	// Reset the program
+	progSingletonLock.Lock()
+	defer progSingletonLock.Unlock()
+	progSingleton = nil
+}
+
+func (e *EbpfProgram) stopImpl() {
+	if e.Manager != nil {
+		_ = e.Manager.Stop(manager.CleanAll)
+	}
```

Comment:
Can you run EbpfProgram with golang race detector?

Reply:
I ran all the tests, the only races were in the test code itself (both the callback and the check were accesing the same function) and fixed them.